compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/Canonicalization.cpp

```diff
@@ -978,16 +978,55 @@ struct ReshapeOpCanon final : OpRewritePattern<mlir::stablehlo::ReshapeOp> {
   }
 };
 
-struct TransposeOpCanon final : OpRewritePattern<mlir::stablehlo::TransposeOp> {
+struct TransposeIsReshape final
+    : OpRewritePattern<mlir::stablehlo::TransposeOp> {
   using OpRewritePattern::OpRewritePattern;
 
   LogicalResult matchAndRewrite(mlir::stablehlo::TransposeOp op,
                                 PatternRewriter &rewriter) const override {
-    // Check if this transpose is a noop and use the operand instead.
-    if (!isIotaRange(op.getPermutation()))
-      return failure();
+    auto input = op.getOperand();
+    auto permutation = op.getPermutation();
 
-    rewriter.replaceOp(op, op.getOperand());
+    if (isIotaRange(permutation)) {
+      rewriter.replaceOp(op, op.getOperand());
+      return success();
+    }
+
+    auto inputTy = dyn_cast<RankedTensorType>(input.getType());
+    if (!inputTy) {
+      return rewriter.notifyMatchFailure(
+          op, "requires input to be of a ranked tensor type");
+    }
+
+    int64_t numDynDims = 0;
+    for (int i = 0; i < inputTy.getRank(); ++i) {
```

Comment:
I would just fail on dynamic shapes for now. `stablehlo.reshape` is pretty different for `tosa.reshape` for dynamic shapes.